src/lib/logical_query_plan/abstract_non_query_node.hpp

```diff
@@ -14,8 +14,11 @@ class AbstractNonQueryNode : public AbstractLQPNode {
   using AbstractLQPNode::AbstractLQPNode;
 
   std::vector<std::shared_ptr<AbstractExpression>> output_expressions() const override;
-  std::shared_ptr<LQPUniqueConstraints> unique_constraints() const override;
-  std::vector<FunctionalDependency> non_trivial_functional_dependencies() const override;
+
+  std::shared_ptr<UniqueColumnCombinations> unique_column_combinations() const override;
```

Comment:
UCCs are often just passed around (they do not change in most cases and are forwarded), whereas the "non-trivial FDs" are usually genuine and not passed around that much _but_ are merged, inflated, whatever (making copies necessary in most cases). So having a pointer for UCCs makes sense IMO to ensure nothing is copied too often. In contrast, we have to make copies of the FD structures anyway. But I'm open if you can convince me to unify the interfaces.

Reply:
Are they passed around or really shared?

Reply:
Passed around, so it would be okay to return the sets when always using references insted of copy assignments. It just makes it harder for anyone to create copies by accident.

Reply:
If it is not shared, why not a unique_ptr? That should avoid any overheads.

Trying to avoid copies makes sense, but (i) clang-tidy finds a lot of these issues and (ii) we already ref-to-vector-passing a lot in Hyrise.

Reply:
So do you have preferences for either unique pointers or ref-to-set? I don't have a strong opinion tbh

Reply:
As we use it widely in the code base, I'd prefer references. But if you think there will be cases where someone will cause copies (usage in plugins, unclear interfaces, ...), I am fine with enforcing it via `unique_ptrs`.

Reply:
When sets are the common interface, I think it should be clear then. For plugins, I guess it's the plugin programmer's responsibility not to do stupid things and they should not interfere with query optimization directly at all. So references it shall be. Gonna ping you when that's done.

Reply:
@Bouncner now only `UniqueColumnCombinations` (a.k.a `std::unordered_set<UniqueColumnCombination>`) is passed around.